File: solutions/kotlin/protein-translation/1/src/main/kotlin/ProteinTranslation.kt
private val PROTEINS_BY_CODON =
    mapOf(
        "AUG" to "Methionine",
        "UUU" to "Phenylalanine",
        "UUC" to "Phenylalanine",
        "UUA" to "Leucine",
        "UUG" to "Leucine",
        "UCU" to "Serine",
        "UCC" to "Serine",
        "UCA" to "Serine",
        "UCG" to "Serine",
        "UAU" to "Tyrosine",
        "UAC" to "Tyrosine",
        "UGU" to "Cysteine",
        "UGC" to "Cysteine",
        "UGG" to "Tryptophan",
    )

private val STOP_CODONS = setOf("UAA", "UAG", "UGA")

fun translate(rna: String?): List<String> = rna
    .orEmpty()
    .chunked(CODON_LENGTH)
    .takeWhile { codon -> codon !in STOP_CODONS }
    .map(::proteinFor)

private fun proteinFor(codon: String): String {
    requireThreeLetterCodon(codon)

    return PROTEINS_BY_CODON[codon] ?: throw IllegalArgumentException(INVALID_CODON)
}

private fun requireThreeLetterCodon(codon: String) {
    require(codon.length == CODON_LENGTH) { INVALID_CODON }
}

private const val CODON_LENGTH = 3
private const val INVALID_CODON = "Invalid codon"

File: solutions/python/protein-translation/1/protein_translation.py
"""Translate RNA codons into proteins."""

PROTEINS_BY_CODON = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan",
}
STOP_CODONS = {"UAA", "UAG", "UGA"}
CODON_LENGTH = 3


def proteins(strand):
    """Return proteins translated from the RNA strand."""
    translated = []

    for codon in _codons(strand):
        if codon in STOP_CODONS:
            break

        translated.append(PROTEINS_BY_CODON[codon])

    return translated


def _codons(strand):
    return (
        strand[index:index + CODON_LENGTH]
        for index in range(0, len(strand), CODON_LENGTH)
    )
